packages/server/postgres/migrations/1635354572171_addIntegrationProviderAndToken.ts

```diff
@@ -0,0 +1,98 @@
+import {Client} from 'pg'
+import getPgConfig from '../getPgConfig'
+
+export async function up() {
+  const client = new Client(getPgConfig())
+  await client.connect()
+  await client.query(`
+  DO $$
+  BEGIN
+    IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'IntegrationProviderTypesEnum') THEN
+      CREATE TYPE "IntegrationProviderTypesEnum" AS ENUM (
+        'GITLAB',
+        'MATTERMOST'
+      );
+    END IF;
+    IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'IntegrationProviderTokenTypeEnum') THEN
+      CREATE TYPE "IntegrationProviderTokenTypeEnum" AS ENUM (
+        'PAT',
+        'OAUTH2',
+        'WEBHOOK'
+      );
+    END IF;
+    IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'IntegrationProviderScopesEnum') THEN
+      CREATE TYPE "IntegrationProviderScopesEnum" AS ENUM (
+        'GLOBAL',
+        'ORG',
+        'TEAM'
+      );
+    END IF;
+    CREATE TABLE IF NOT EXISTS "IntegrationProvider" (
+      "id" INT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+      "type" "IntegrationProviderTypesEnum" NOT NULL,
+      "tokenType" "IntegrationProviderTokenTypeEnum" NOT NULL,
+      "scope" "IntegrationProviderScopesEnum" NOT NULL,
+      "scopeGlobal" BOOLEAN GENERATED ALWAYS AS (
+        CASE
+          WHEN "scope" = 'GLOBAL' THEN TRUE
+          ELSE NULL
+        END
+      ) STORED,
+      "orgId" VARCHAR(100),
+      "teamId" VARCHAR(100),
+      "isActive" BOOLEAN DEFAULT TRUE NOT NULL,
+      "name" VARCHAR(250) NOT NULL,
+      "serverBaseUri" VARCHAR(2083) NOT NULL,
+      "oauthScopes" VARCHAR(100)[],
+      "oauthClientId" VARCHAR(2600),
+      "oauthClientSecret" VARCHAR(2600),
+      "createdAt" TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP NOT NULL,
+      "updatedAt" TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP NOT NULL,
+      UNIQUE("scopeGlobal", "type"),
+      CONSTRAINT global_provider_must_be_oauth2 CHECK (
+        "scopeGlobal" IS NULL OR ("scopeGlobal" = TRUE AND "tokenType" = 'OAUTH2')
+      )
+    );
+    CREATE INDEX IF NOT EXISTS "idx_IntegrationProvider_typeAndScope"
+      ON "IntegrationProvider"("type", "scope");
+    CREATE INDEX IF NOT EXISTS "idx_IntegrationProvider_orgId" ON "IntegrationProvider"("orgId");
+    CREATE INDEX IF NOT EXISTS "idx_IntegrationProvider_teamId" ON "IntegrationProvider"("teamId");
+    CREATE TABLE IF NOT EXISTS "IntegrationToken" (
+      "teamId" VARCHAR(100) NOT NULL,
```

Comment:
Some thoughts on whether or not we should have the IntegrationToken per User, Team, Provider (1, current solution) or just User, Provider and then a separate relation TeamMemberIntegration (2).

1) **Pro**
  * easy, it's what we have now
  * no joins
**Con**
  * updates to the token need to be propagated to all rows
2) **Pro**
  * it would simplify refreshing a token and similar operations on the token itself, because it would not require to update multiple rows.
  * integration specific settings could be stored on TeamMemberIntegration and thus could also be stored without having a token
 **Con**
  * additional join for retrieving the token
  * the key User,IntegrationProvider might not work. While I think it is unlikely that someone wants to have access to Jira with 2 different accounts in 1 team, the likelyhood increases when it's spread over multiple teams. This is probably still not a use case, only wanted to mention it.